File: src/gravel.ts
export interface GravelInput {
  shape: 'rectangular' | 'circular';
  /** ft (imperial) or m (metric) */
  length?: number;
  width?: number;
  radius?: number;
  /** ft (imperial) or m (metric) */
  depth: number;
  pricePerTon?: number;
  units: 'imperial' | 'metric';
}

export interface GravelResult {
  /** yd³ (imperial) or m³ (metric) */
  volume: number;
  /** Converted to the other unit */
  alternativeVolume: number;
  tonnage: number;
  bags50lb: number;
  estimatedCost: number;
}

/** Calculate gravel/aggregate volume, weight, and cost. */
export function calculateGravel(input: GravelInput): GravelResult {
  const { shape, depth, pricePerTon = 0, units } = input;

  let volume: number;
  if (shape === 'circular') {
    const r = input.radius ?? 0;
    volume = units === 'imperial'
      ? (Math.PI * r * r * depth) / 27
      : Math.PI * r * r * depth;
  } else {
    const l = input.length ?? 0;
    const w = input.width ?? 0;
    volume = units === 'imperial'
      ? (l * w * depth) / 27
      : l * w * depth;
  }

  const alternativeVolume = units === 'imperial'
    ? volume * 0.764555
    : volume * 1.30795;

  const tonnage = units === 'imperial' ? volume * 1.4 : volume * 1.5;
  const bags50lb = units === 'imperial'
    ? Math.ceil(tonnage * 40)
    : Math.ceil(tonnage * 44);

  const estimatedCost = tonnage * pricePerTon;

  return { volume, alternativeVolume, tonnage, bags50lb, estimatedCost };
}

File: python/construction_calculators/gravel.py
"""Gravel and aggregate volume, weight, and cost calculator."""

import math
from dataclasses import dataclass, field
from typing import Literal


@dataclass
class GravelInput:
    """Input parameters for gravel calculation.

    Imperial: dimensions in feet, depth in feet.
    Metric: dimensions in meters.
    """

    shape: Literal["rectangular", "circular"]
    depth: float
    units: Literal["imperial", "metric"]
    length: float = 0.0
    width: float = 0.0
    radius: float = 0.0
    price_per_ton: float = 0.0


@dataclass
class GravelResult:
    """Result of gravel calculation."""

    volume: float
    """Volume in yd³ (imperial) or m³ (metric)."""
    alternative_volume: float
    """Volume converted to the other unit."""
    tonnage: float
    """Weight in short tons (imperial) or metric tons."""
    bags_50lb: int
    """Equivalent number of 50 lb bags."""
    estimated_cost: float
    """Estimated cost based on price_per_ton."""


def calculate_gravel(input: GravelInput) -> GravelResult:
    """Calculate gravel/aggregate volume, weight, and cost.

    Args:
        input: Gravel calculation parameters.

    Returns:
        GravelResult with volume, tonnage, bag count, and cost.

    Examples:
        >>> from construction_calculators.gravel import GravelInput, calculate_gravel
        >>> inp = GravelInput(shape="rectangular", length=10, width=10, depth=0.25, units="imperial", price_per_ton=35)
        >>> result = calculate_gravel(inp)
        >>> print(f"{result.volume:.2f} yd³, {result.tonnage:.2f} tons")
        0.93 yd³, 1.30 tons
    """
    shape = input.shape
    depth = input.depth
    units = input.units

    if shape == "circular":
        r = input.radius
        if units == "imperial":
            volume = (math.pi * r * r * depth) / 27
        else:
            volume = math.pi * r * r * depth
    else:
        l = input.length
        w = input.width
        if units == "imperial":
            volume = (l * w * depth) / 27
        else:
            volume = l * w * depth

    alternative_volume = volume * 0.764555 if units == "imperial" else volume * 1.30795

    tonnage = volume * 1.4 if units == "imperial" else volume * 1.5
    bags_50lb = (
        math.ceil(tonnage * 40) if units == "imperial" else math.ceil(tonnage * 44)
    )
    estimated_cost = tonnage * input.price_per_ton

    return GravelResult(
        volume=volume,
        alternative_volume=alternative_volume,
        tonnage=tonnage,
        bags_50lb=bags_50lb,
        estimated_cost=estimated_cost,
    )
